In [313]:
import pandas as pd
import numpy as np
import seaborn as sb
df_attacks = pd.read_csv("DATA/attacks.csv",encoding = "ISO-8859-1")
import re
import sys
sys.path.append('../')

import src.cleaning_functions as cf

MIS 3 HIPÓTESIS SON:
- En 2018 hubo más ataques de tiburones tigre en el hemisferio sur que en 1990.
- En México hay más ataques de tiburones blancos que en USA.
- La mayoría de las víctimas que practican surf son mujeres.

Por tanto, para verificar o refutar las 3 hipótesis planteadas, voy a utilizar las siguientes columnas:
- Year
- Country
- Activity
- Sex
- Species

Voy a proceder a eliminar las columnas del dataframe que, de primeras, no me van a ser de utilidad:

In [314]:
df_attacks.drop(["Case Number","pdf","href formula","href","Case Number.1","Case Number.2","original order","Unnamed: 22","Unnamed: 23"], axis=1, inplace=True)

Ahora la forma de mi df es:

In [315]:
df_attacks.shape

(25723, 15)

Borro filas duplicadas y comprobamos que hemos pasado de 25723 a 6305 filas.

In [316]:
borrado=df_attacks.drop_duplicates()
borrado.shape


(6302, 15)

Borro las filas que sean enteras NaN. En este caso, tras comprobar el shape, veo que solo es una.

In [317]:
borrado.dropna(how="all", inplace=True)
borrado.shape

(6301, 15)

Con ".info" veo qué tipos de datos tengo y veo que todos son categóricos:

In [318]:
borrado.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6301 entries, 0 to 6301
Data columns (total 15 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Date                    6301 non-null   object 
 1   Year                    6299 non-null   float64
 2   Type                    6297 non-null   object 
 3   Country                 6251 non-null   object 
 4   Area                    5846 non-null   object 
 5   Location                5761 non-null   object 
 6   Activity                5757 non-null   object 
 7   Name                    6091 non-null   object 
 8   Sex                     5736 non-null   object 
 9   Age                     3471 non-null   object 
 10  Injury                  6273 non-null   object 
 11  Fatal (Y/N)             5762 non-null   object 
 12  Time                    2948 non-null   object 
 13  Species                 3463 non-null   object 
 14  Investigator or Source  6284 non-null   

Remplazo los "nan" por "Unkown" para cada una de las columnas:

In [319]:
borrado["Date"].fillna(("Unknown"), inplace=True)
borrado["Year"].fillna(("Unknown"), inplace=True)
borrado["Type"].fillna(("Invalid"), inplace=True)
borrado["Country"].fillna(("Unknown"),inplace=True)
borrado["Area"].fillna(("Unknown"), inplace=True)
borrado["Location"].fillna(("Unknown"), inplace=True)
borrado["Activity"].fillna(("Unknown"), inplace=True)
borrado["Name"].fillna(("Unknown"), inplace=True)
borrado["Sex "].fillna(("Unknown"),inplace=True)
borrado["Age"].fillna(("Unknown"), inplace=True)
borrado["Injury"].fillna(("Unknown"), inplace=True)
borrado["Fatal (Y/N)"].fillna(("Unknown"), inplace=True)
borrado["Time"].fillna(("Unknown"), inplace=True)
borrado["Species "].fillna(("Unknow"), inplace=True)
borrado["Investigator or Source"].fillna(("Unknown"),inplace=True)



Voy a empezar a armonizar los elementos de cada columna.


Empiezo por la columna de "Sex".
- Compruebo los valores únicos de la columna "Sex".
- Sustituyo la "M con espacio" por la "M normal". El resto las replazo por "unkonwn" y me quedo únicamente por M y F. 
- Compruebo de nuevo los valores únicos de "Sex" y veo que se han reducido a:"F","M","nan" y "Unknown".

In [320]:
borrado["Sex "].unique()

df2=borrado.replace({'M ':'M','lli':'Unknown','N':'Unknown','.':'Unknown'})

df2["Sex "].unique()


array(['F', 'M', 'Unknown'], dtype=object)

Continúo con la columa "Species":

In [321]:
especies=df2["Species "].unique()

In [322]:
df2["Species "]= df2["Species "].str.replace(".*[Ww](hite|HITE).*","White shark",regex=True)
df2["Species "]= df2["Species "].str.replace(".*[Tt](iger|IGER).*","Tiger shark",regex=True)
df2["Species "]= df2["Species "].str.replace(".*[Ll](emon|EMON).*","Lemon shark",regex=True)
df2["Species "]= df2["Species "].str.replace(".*[Hh](ammerhead|AMMERHEAD).*","Hammerhead shark",regex=True)
df2["Species "]= df2["Species "].str.replace(".*[Bb](ull|ULL).*","Bull shark",regex=True)
df2["Species "]= df2["Species "].str.replace(".*[Bb](lue|LUE).*","Blue shark",regex=True)
df2["Species "]= df2["Species "].str.replace(".*[Ss](ilvertip|ILVERTIP).*","Silvertip shark",regex=True)
df2["Species "]= df2["Species "].str.replace(".*[Nn](urse|URSE).*","Nurse shark",regex=True)
df2["Species "]= df2["Species "].str.replace(".*[Ww](haler|HALER).*","Whaler shark",regex=True)
df2["Species "]= df2["Species "].str.replace(".*[Bb](lacktip|LACKTIP).*","Blacktip shark",regex=True)
df2["Species "]= df2["Species "].str.replace(".*[MM](ako|AKO).*","Mako shark",regex=True)
df2["Species "]= df2["Species "].str.replace(".*[Ss](and|AMD).*","Sand shark",regex=True)
df2["Species "]= df2["Species "].str.replace(".*[Ww](obbegong|OBBEGONG).*","Wobbegong shark",regex=True)
df2["Species "]= df2["Species "].str.replace(".*[Gg](alapagos|ALAPAGOS).*","Galapagos shark",regex=True)
df2["Species "]= df2["Species "].str.replace(".*[Gg](rey|REY).*","Grey shark",regex=True)
df2["Species "]= df2["Species "].str.replace(".*[Ll](eopard|EOPARD).*","Leopard shark",regex=True)
df2["Species "]= df2["Species "].str.replace(".*[Zz](ambesi|AMBESI).*","Zambesi shark",regex=True)
df2["Species "]= df2["Species "].str.replace(".*[Bb](lacktail|LACKTAIL).*","Blacktail shark",regex=True)
df2["Species "]= df2["Species "].str.replace(".*[Rr](ed|ED).*","Red shark",regex=True)
df2["Species "]= df2["Species "].str.replace(".*[Dd](usky|USKY).*","Dusky shark",regex=True)
df2["Species "]= df2["Species "].str.replace(".*[Rr](aggedtooth|AGGEDTOOTH).*","Raggedtooth shark",regex=True)
df2["Species "]= df2["Species "].str.replace(".*[Ss](pinner|PINNER).*","Spinner shark",regex=True)
df2["Species "]= df2["Species "].str.replace(".*[Cc](ow|OW).*","Cow shark",regex=True)
df2["Species "]= df2["Species "].str.replace(".*[Pp](orbeagle|ORBEAGLE).*","Porbeagle shark",regex=True)
#df2["Species "]= df2["Species "].str.replace("[^.?!]*Bronze[^.?!]*","Bronze whaler shark",regex=True) #este no sale cuando filtro por "Bronze whaler shark". Ahora me desaparece Bronze directamente...
df2["Species "]= df2["Species "].str.replace(".*[Cc](aribbean|ARIBBEAN).*","Caribbean reef shark",regex=True)
df2["Species "]= df2["Species "].str.replace(".*[Ss](and|AND).*","Sandbar shark",regex=True)
df2["Species "]= df2["Species "].str.replace(".*[Ss](ilky|ILKY).*","Silky shark",regex=True)
df2["Species "]= df2["Species "].str.replace(".*[Zz](ambezi|AMBEZI).*","Zambezi shark",regex=True)
df2["Species "]= df2["Species "].str.replace(".*[Ss](evengill|EVENGILL).*","Sevengill shark",regex=True)
df2["Species "]= df2["Species "].str.replace(".*[Cc](opper|OPPER).*","Copper shark",regex=True)
df2["Species "]= df2["Species "].str.replace(".*[Aa](ngel|NGEL)\s","Angel shark",regex=True) #lo he puesto diferente para que no pille "los ángeles" aún así parece que no ha funcionado.
df2["Species "]= df2["Species "].str.replace(".*[Ss](almon|ALMON).*","Salmon shark",regex=True)
df2["Species "]= df2["Species "].str.replace(".*[Gg](oblin|OBLIN).*","Goblin shark",regex=True)
df2["Species "]= df2["Species "].str.replace(".*[Tt](hresher|HRESHER).*","Thresher shark",regex=True)
df2["Species "]= df2["Species "].str.replace(".*[Dd](ogfish|OGFISH).*","Dogfish shark",regex=True)
df2["Species "]= df2["Species "].str.replace("[^.?!]*involvement[^.?!]*","Involvement not confirmed",regex=True)#no funciona



Saco los valores únicos tras la agruapación y los convierto en lista:

In [323]:
especieslimpio=df2["Species "].unique()
especieslimpio
len(especieslimpio)
listaespecieslimpio=list(especieslimpio)
#print(listaespecieslimpio)

Usando "value.counts", obtengo la frecuencia de cada uno de los valores únicos que he obtenido en el paso anterior y la meto en una variable lista:

In [324]:
frecuencia_especies=list(df2["Species "].value_counts(sort=False))

Ahora me creo una lista solo con las frecuencias que son igual a 1. Entiendo que serán los valores que quedan sin armonizar y que quiero meter bajo el nombre genérico "Unkown".

In [325]:
lista2=[]
for i in range(len(frecuencia_especies)):
    if frecuencia_especies[i]==1:
        lista2.append(frecuencia_especies[i])
#print(lista2)
#len(frecuencia_especies)    

Utilizando la función zip con la lista de los valores únicos de la columna Especies y la lista de las frecuencias de las especies, hago un diccionario en el que figura cada uno de los elementos de la columna especie como key y su frecuencia como value.

In [326]:
dicc_ = {}
for especieslimpio,frecuencia_especies in zip(especieslimpio,frecuencia_especies):
    dicc_[especieslimpio] = frecuencia_especies

In [327]:
#print(dicc_ )

El siguiente paso es filtrar aquellos pares cuyo value sea 1. Luego, obtengo las keys y la convierto en lista. Entonces, ya tendría todos los elementos que quiero renombrar con "Unkown".

In [328]:
especiesunkown=dict(filter(lambda x: x[1] == 1, dicc_ .items()))
type(especiesunkown)
#print(especiesunkown)
listaespeciesunkown=list(especiesunkown.keys())
#print((listaespeciesunkown))
#type(listaespeciesunkown)


Construyo una función que recibe dos listas de tamaños diferentes. En este caso, la lista de valores únicos de la columna Especies (la mayor) y otra lista con los valores que quiero sustituir por Unkown (la menor). La función me tiene que devolver la lista de los valores únicos con los elementos de la lista menor reemplazados por "unkown".

In [329]:
#def sustituir(s,b):
    #for i,p in enumerate(s):
       # if p in b:
            #b[i]="unkown"
   # return b

Con la lista que me devuelve la función, trataría de machacar la columna Especies ya existente. No sé si se podría utilizar .apply o si sería mejor crear una columna nueva y asignarles los valores de la lista resultante de la función para ya luego borrar la columna original de Especies.

In [348]:
df2["Species"]=df2["Species "].apply(cf.sustituir) #SOS!!!!


AttributeError: module 'src.cleaning_functions' has no attribute 'sustituir'

In [151]:
list(df2["Activity"].unique()) # Usar redex. Por ejemplo: donde me encuentre swim sustituye por swimming y así voy creando menos categorías usando replace el metodo de redex por lo otro. str.replace. Las cosas raras que no pueda incluir en ninguna categoría, las pongo como "others"
df2["Activity"].value_counts().head(50)

Surfing                           971
Swimming                          869
Unknown                           550
Fishing                           431
Spearfishing                      332
Bathing                           162
Wading                            149
Diving                            127
Standing                           99
Snorkeling                         89
Scuba diving                       76
Body boarding                      61
Body surfing                       49
Swimming                           47
Kayaking                           33
Fell overboard                     32
Treading water                     32
Pearl diving                       32
Boogie boarding                    29
Free diving                        29
Windsurfing                        19
Walking                            17
Boogie Boarding                    16
Shark fishing                      15
Floating                           14
Canoeing                           13
Fishing     

Obtengo los valores únicos de la columa Países:

In [126]:
(df2["Country"].unique())

array(['USA', 'AUSTRALIA', 'MEXICO', 'BRAZIL', 'ENGLAND', 'SOUTH AFRICA',
       'THAILAND', 'COSTA RICA', 'MALDIVES', 'BAHAMAS', 'NEW CALEDONIA',
       'ECUADOR', 'MALAYSIA', 'LIBYA', 'Unknown', 'CUBA', 'MAURITIUS',
       'NEW ZEALAND', 'SPAIN', 'SAMOA', 'SOLOMON ISLANDS', 'JAPAN',
       'EGYPT', 'ST HELENA, British overseas territory', 'COMOROS',
       'REUNION', 'FRENCH POLYNESIA', 'UNITED KINGDOM',
       'UNITED ARAB EMIRATES', 'PHILIPPINES', 'INDONESIA', 'CHINA',
       'COLUMBIA', 'CAPE VERDE', 'Fiji', 'DOMINICAN REPUBLIC',
       'CAYMAN ISLANDS', 'ARUBA', 'MOZAMBIQUE', 'FIJI', 'PUERTO RICO',
       'ITALY', 'ATLANTIC OCEAN', 'GREECE', 'ST. MARTIN', 'FRANCE',
       'PAPUA NEW GUINEA', 'TRINIDAD & TOBAGO', 'KIRIBATI', 'ISRAEL',
       'DIEGO GARCIA', 'TAIWAN', 'JAMAICA', 'PALESTINIAN TERRITORIES',
       'GUAM', 'SEYCHELLES', 'BELIZE', 'NIGERIA', 'TONGA', 'SCOTLAND',
       'CANADA', 'CROATIA', 'SAUDI ARABIA', 'CHILE', 'ANTIGUA', 'KENYA',
       'RUSSIA', 'TURKS & CAICOS', 

Ahora usando ".apply" importo y aplico la función que he creado en "Cleaning_functions.py"para determinar si cada país de la columna Country pertenece al Hemisferio Norte o al Hemisferio Sur y veo que efectivamente se crea otra columna llamada Hemisferio.

In [338]:
df2["Hemisphere"] = df2["Country"].apply(cf.hemisferio) 
#no me pilla el import de mi hoja de funciones de visual code.

In [339]:
df2.head()

Date    Year        Type    Country             Area  \
0  25-Jun-2018  2018.0     Boating        USA       California   
1  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex       Age  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F       57   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely     F       11   
2                    Habush, Oahu      Surfing      John Denges    M       48   
3              Arrawarra Headland      Surfing             male    M  Unknown   
4                        La Ticla  Free diving   Gustavo Ramos     M  Unknown   

                                              Injury Fatal (Y/N)  \
0  No injury to occupant, outrigger canoe and pad...     Unknown   
1                         Minor injury to left thigh     Unknown   
2       Injury to left lower leg from surfboard skeg     Unknown   
3                          Minor injury to lower leg     Unknown   
4  Lacerations to leg & hand shark PROVOKED INCIDENT     Unknown   

            Time     Species           Investigator or Source  \
0          18h00  White shark                R. Collier, GSAF   
1  14h00  -15h00       Unknow  K.McMurray, TrackingSharks.com   
2          07h45       Unknow  K.McMurray, TrackingSharks.com   
3        Unknown    2 m shark                  B. Myatt, GSAF   
4        Unknown  Tiger shark                       A .Kipper   

         Hemisphere  
0  Hemisferio norte  
1  Hemisferio norte  
2  Hemisferio norte  
3    Hemisferio sur  
4  Hemisferio norte

In [129]:
df2["Area"].unique()
df2["Area"].value_counts()

Florida                                 1037
New South Wales                          486
Unknown                                  455
Queensland                               311
Hawaii                                   297
                                        ... 
Ysabel Island                              1
 Lau Province                              1
South Coast, East New Britain              1
Between Southampton & Canary Islands       1
Moala Island                               1
Name: Area, Length: 826, dtype: int64

Veo cuáles son los valores únicos de la columna "Type" y luego hago replace para armonizar sus elementos y reducirlos a 5:

In [130]:
#tipo=df2["Type"].unique()
df2["Type"].value_counts()

Unprovoked      4594
Provoked         574
Invalid          553
Boating          341
Sea Disaster     239
Name: Type, dtype: int64

In [131]:
df2.replace({"Questionable":"Invalid","Boat":"Boating","Boatomg":"Boating"},inplace=True)

In [132]:
df2["Type"].value_counts()

Unprovoked      4594
Provoked         574
Invalid          553
Boating          341
Sea Disaster     239
Name: Type, dtype: int64

Regularizo la columna nombre: veo que, al hacer los valores únicos, hay muchos valores que no son nombres reales y que se repiten mucho. He decidido que todos aquellos valores que se repitan más de 3 veces, los voy a sustituir por "Unkown"

In [141]:
nombres=list(df2["Name"].unique())
#print(nombres)
frecuencianombres=list(df2["Name"].value_counts(sort=False)) #Podemos incluir el parámetro "Sort=False" para que no me ordene la frecuenecia con la que aparecen los valores únicos de forma descendente y lo deje en el miso orden que aparecen en el dataframe.
#print(frecuencianombres)


In [134]:
dicc_nombre = {}
for nombres,frecuencianombres in zip(nombres,frecuencianombres):
    dicc_nombre[nombres] = frecuencianombres

In [143]:
#print(dicc_nombre)

In [149]:
nombresunkown=dict(filter(lambda x: x[1] > 3, dicc_nombre.items()))
type(nombresunkown)
#rint(nombresunkown)
listanombresunkown=list(nombresunkown.keys())
print((listanombresunkown))
#type(listanombresunkown)


['male', 'female', 'Unknown', 'girl', 'child', 'boy', '2 males', 'Unidentified', 'Anonymous', 'black male', '2 fishermen', 'a soldier', 'boat', 'fisherman', 'a pearl diver', 'sailor', 'a sailor', 'males', 'a native', 'Arab boy']


Ahora utilizaría la función "sustituir" creada anteriormente para otra columna y con la lista que me devuelve la función, trataría de machacar la columna "Name" ya existente. No sé si se podría utilizar .apply o si sería mejor crear una columna nueva y asignarles los valores de la lista resultante de la función para ya luego borrar la columna original de Name. **Luego podría quizás depurar un poco más eseta columna con Regex quitando sustituyendo por "Unkown" aquellos elementos que contengan "male".

Ahora voy a ver los valores únicos de la columna "Fatal (Y/N)", dejando únicamente 3:Y,N y Unkown.

In [168]:
df2["Fatal (Y/N)"].value_counts()

Unknown    4903
Y          1389
N             7
Unkown        2
Name: Fatal (Y/N), dtype: int64

In [171]:
df2.replace({"UNKNOWN":"Unknown"," N":"N","y":"Y","M":"Unknown","2017":"Unknown","N ":"Unknown","Unkown":"Unknown" },inplace=True)

In [172]:
df2["Fatal (Y/N)"].value_counts()

Unknown    4905
Y          1389
N             7
Name: Fatal (Y/N), dtype: int64

Busco los valores únicos de la columna "Injury" y, mediante el uso de Regex, sustituyo los valores armonizados para disminuir el número de valores únicos:

In [193]:
df2["Injury"].unique()


array(['No injury', 'Minor injury to left thigh',
       'Injury to left lower leg from surfboard skeg', ...,
       'FATAL, leg stripped of flesh  ',
       'FATAL, knocked overboard by tail of shark & carried off by shark ',
       'FATAL. "Shark bit him in half, carrying away the lower extremities" '],
      dtype=object)

In [258]:
df2["Injury"]=df2["Injury"].str.replace(".*[Nn](o|O)\s*[Ii](njury|NJURY).*","No injury",regex=True)
df2["Injury"]=df2["Injury"].str.replace(".*[Mm](inor|INOR)\s*[Ii](njury|NJURY|njuries|NJURIES|ed|ED).*","Minor Injury",regex=True)
df2["Injury"]=df2["Injury"].str.replace(".*[Hh](and|AND).*","Hand injury",regex=True)
df2["Injury"]=df2["Injury"].str.replace(".*[Ll](eg|EG).*","Leg injury",regex=True)
df2["Injury"]=df2["Injury"].str.replace(".*[Ll](aceration|ACERATION).*","Laceration",regex=True)
df2["Injury"]=df2["Injury"].str.replace(".*[Ff](atal|ATAL).*","Fatal",regex=True)
df2["Injury"]=df2["Injury"].str.replace(".*[Ff](oot|OOT).*","Foot Injury",regex=True)
df2["Injury"]=df2["Injury"].str.replace(".*[Aa](nkle|NKLE).*","Ankle Injury",regex=True)
df2["Injury"]=df2["Injury"].str.replace(".*[Ss](houlder|HOULDER).*","Shoulder Injury",regex=True)
df2["Injury"]=df2["Injury"].str.replace(".*[Aa](brasion|BRASION).*","Abrasion Injury",regex=True)
df2["Injury"]=df2["Injury"].str.replace(".*[Cc](alf|ALF).*","Calf Injury",regex=True)
df2["Injury"]=df2["Injury"].str.replace(".*[Tt](orso|ORSO).*","Torso Injury",regex=True)
df2["Injury"]=df2["Injury"].str.replace(".*[Ff](eet|EET).*","Feet Injury",regex=True)
df2["Injury"]=df2["Injury"].str.replace(".*[Tt](high|HIGH).*","Thigh Injury",regex=True)
df2["Injury"]=df2["Injury"].str.replace(".*[Aa](rm|RM)\s*[Ii](njury|NJURY).*","Arm injury",regex=True)
df2["Injury"]=df2["Injury"].str.replace(".*[Pp](uncture|UNCTURE)\s*[Ww](ounds|OUNDS).*","Puncture wounds",regex=True)
df2["Injury"]=df2["Injury"].str.replace(".*[Kk](nee|NEE).*","Knee Injury",regex=True)
df2["Injury"]=df2["Injury"].str.replace(".*[Cc](hest|HEST).*","Chest Injury",regex=True)
df2["Injury"]=df2["Injury"].str.replace(".*[Ee](lbow|LBOW).*","Elbow Injury",regex=True)

Aquí tendría que hacer la función remplazar para esta columna también...

En resumen, he limpiado los datos de las siguientes columnas:
- Year:acaban en punto 0, quizás podría quitarse.	
- Type: he dejado 5 valores únicos(Unprovoked, provoked, Invalid, Boating, Sea Disaster).	
- Country: la he utilizado para crear la columna "Hemisphere".		
- Activity: Bea.
- Name: he dejado aquellos elementos que se repiten menos de 4 veces e importo con la función "sustituir). Quizás luego depure con Regex.	
- Sex: he dejado 3 valores únicos(F/M/Unknown).	
- Injury: regularizado con Regex e importado con una función(sustituir).
- Fatal (Y/N): he dejado 3 valores únicos(Y/N/Unknown).	
- Species:regularizado con Regex e importado con una función(sustituir).
- Hemisphere: he creado esta columna con la función hemisferio.

He dejado tal cual estaban las siguientes columnas:
- Investigator or Source.
- Area.
- Age.
- Time.
- Location
- Date
